File: src/main/java/de/twometer/orion/util/Cache.java
package de.twometer.orion.util;

import java.util.HashMap;
import java.util.Map;

public abstract class Cache<ID, T> {

    private final Map<ID, T> cache = new HashMap<>();

    protected abstract T create(ID id) throws Exception;

    public T get(ID id) {
        var result = cache.get(id);
        if (result == null) {
            try {
                result = create(id);
            } catch (Exception e) {
                CrashHandler.fatal(e);
                return null;
            }
            cache.put(id, result);
        }
        return result;
    }

}

File: neko-engine/src/main/kotlin/de/twometer/neko/util/Cache.kt
package de.twometer.neko.util

abstract class Cache<K, V> {

    private val map = HashMap<K, V>()

    protected abstract fun create(key: K): V

    fun preload(key: K) {
        if (!map.containsKey(key))
            map[key] = create(key)
    }

    fun get(key: K): V {
        var result = map[key]
        if (result == null) {
            result = create(key)
            map[key] = result
        }
        return result!!
    }

    fun map() = map

}